libs/aws/langchain_aws/chat_models/bedrock_converse.py

```diff
@@ -1750,3 +1796,61 @@ def _is_cache_point(cache_point: Any) -> bool:
         and "cachePoint" in cache_point
         and cache_point.get("cachePoint").get("type") is not None
     )
+
+
+def _has_tool_use_or_result_blocks(messages: List[Dict[str, Any]]) -> bool:
+    """Check if messages contain toolUse or toolResult blocks."""
+    for message in messages:
+        for block in message.get("content", []):
+            if "toolUse" in block or "toolResult" in block:
+                return True
+    return False
+
+
+def _convert_tool_blocks_to_text(
+    messages: List[Dict[str, Any]],
+) -> List[Dict[str, Any]]:
+    """Convert toolUse and toolResult blocks to text blocks preserving only necessary content."""
+    converted_messages = []
+
+    for message in messages:
+        converted_message = {"role": message["role"], "content": []}
+
+        for block in message.get("content", []):
+            if "toolUse" in block:
+                # convert toolUse to simple text
+                tool_use = block["toolUse"]
+                tool_name = tool_use.get("name", "function")
+                tool_inputs = tool_use.get("input", {})
+
+                # format function call description
+                if tool_inputs:
+                    tool_text = f"[Called {tool_name} with parameters: {json.dumps(tool_inputs)}]"
+                else:
+                    tool_text = f"[Called {tool_name}]"
+
+                converted_message["content"].append({"text": tool_text})
+
+            elif "toolResult" in block:
+                # convert toolResult to indicate it's tool output without exposing internal details
+                tool_result = block["toolResult"]
+
+                result_content = ""
+                for content_block in tool_result.get("content", []):
+                    if "text" in content_block:
+                        result_content += content_block["text"]
+                    elif "json" in content_block:
+                        result_content += json.dumps(content_block["json"])
+                    # skip other internal content types
```

Comment:
Bit of a nit, but when concatenating strings in a loop like this, it's significantly more efficient to collect them in a list, then use `.join` and build out the final string at the end. 

(Ref: https://waymoot.org/home/python_string/)

```suggestion
                content_parts = []
                for content_block in tool_result.get("content", []):
                    if "text" in content_block:
                        content_parts.append(content_block["text"])
                    elif "json" in content_block:
                        content_parts.append(json.dumps(content_block["json"]))
                    # skip other internal content types
                result_content = "".join(content_parts)
```

Reply:
Thanks for the reference!